### PR TITLE
refactor: clean up interface `QueryLastCommittedPublicRand()`

### PR DESCRIPTION
**Pull request by bap2pecs**
## Summary

#407 

## Test Plan

```
make lint
```
wait for CI

### CODE REVIEW

File: types/pub_rand_commit.go [hunk 1]
```diff
@@ -7,8 +7,9 @@ import (
 )
 
 type PubRandCommit struct {
-	NumPubRand uint64 `json:"num_pub_rand"`
-	Commitment []byte `json:"commitment"`
+	StartHeight uint64 `json:"start_height"`
```
Comment (SebastianElvis) [inline types/pub_rand_commit.go:10]: Why is this added back (although I do prefer to have it)?

File: clientcontroller/babylon/consumer.go [hunk 2]
```diff
@@ -298,15 +297,16 @@ func (bc *BabylonConsumerController) QueryLastCommittedPublicRand(fpPk *btcec.Pu
 		return nil, fmt.Errorf("failed to query committed public randomness: %w", err)
 	}
 
-	commitMap := make(map[uint64]*types.PubRandCommit)
-	for height, commit := range res.PubRandCommitMap {
-		commitMap[height] = &types.PubRandCommit{
-			NumPubRand: commit.NumPubRand,
-			Commitment: commit.Commitment,
+	var commit types.PubRandCommit
+	for height, commitRes := range res.PubRandCommitMap {
```
Comment (SebastianElvis) [inline clientcontroller/babylon/consumer.go:301]: Let's play safe here: if `len(res.PubRandCommitMap) == 1`, do the following, otherwise return `nil, nil`.

Currently `commit` might be given an empty (but not nil) value, not 100% sure. Maybe this is also why e2e fails?

File: clientcontroller/cosmwasm/consumer.go [hunk 3]
```diff
@@ -421,16 +422,17 @@ func (wc *CosmwasmConsumerController) QueryLastCommittedPublicRand(fpPk *btcec.P
...
 
 	// Convert the response to the expected map format
-	commitMap := make(map[uint64]*fptypes.PubRandCommit)
-	for _, commit := range commits {
-		commitCopy := commit // create a copy to avoid referencing the loop variable
-		commitMap[commit.StartHeight] = &fptypes.PubRandCommit{
-			NumPubRand: commitCopy.NumPubRand,
-			Commitment: commitCopy.Commitment,
+	var commit *fptypes.PubRandCommit
+	for _, commitRes := range commits {
```
Comment (SebastianElvis) [inline clientcontroller/cosmwasm/consumer.go:426]: Same here

File: types/pub_rand_commit.go (hunk 1 above)
Comment (bap2pecs) [reply types/pub_rand_commit.go:10]: b/c now we don't need both
```
type PubRandCommit struct {
	NumPubRand uint64 `json:"num_pub_rand"`
	Commitment []byte `json:"commitment"`
}
```

and

```
type LastPubRandCommitResponse struct {
	StartHeight uint64 `json:"start_height"`
	NumPubRand  uint64 `json:"num_pub_rand"`
	Commitment  []byte `json:"commitment"`
}
```

File: clientcontroller/babylon/consumer.go (hunk 2 above)
Comment (lesterli) [reply clientcontroller/babylon/consumer.go:301]: Yes, indeed the reason is that `commit` is an empty value but not `nil`.

Comment (bap2pecs) [reply clientcontroller/babylon/consumer.go:305]: this is probably a better condition:

```
	if len(res.PubRandCommitMap) > 1 {
		return nil, fmt.Errorf("expected length to be 1, but got :%d", len(res.PubRandCommitMap))
	}

	var commit *types.PubRandCommit
	for height, commitRes := range res.PubRandCommitMap {
		commit = &types.PubRandCommit{
			StartHeight: height,
			NumPubRand:  commitRes.NumPubRand,
			Commitment:  commitRes.Commitment,
		}
	}

	return commit, nil
```

Comment (SebastianElvis) [reply clientcontroller/babylon/consumer.go:305]: Shouldn't we still check whether `len(res.PubRandCommitMap) == 0` and returns `nil, nil` if this is true?

Comment (bap2pecs) [reply clientcontroller/babylon/consumer.go:305]: wait I think the above code will return nil nil when it's 0 length. is it not true?

Comment (SebastianElvis) [reply clientcontroller/babylon/consumer.go:305]: Right, `commit` is now a pointer so in theory it will be `nil` if `res` is empty. Personally I feel it's a bit unintuitive so prefer `var commit *types.PubRandCommit = nil`, but no strong opinion, your call

**Review by SebastianElvis — APPROVED** [review]
Great work! Would be good to get feedback from @gitferry as well since this is an interface level change

File: finality-provider/service/fp_instance.go [hunk 4]
```diff
@@ -891,31 +891,29 @@ func (fp *FinalityProviderInstance) getPollerStartingHeight() (uint64, error) {
...
 func (fp *FinalityProviderInstance) GetLastCommittedHeight() (uint64, error) {
-	pubRandCommitMap, err := fp.lastCommittedPublicRandWithRetry(1)
+	pubRandCommit, err := fp.lastCommittedPublicRandWithRetry()
 	if err != nil {
 		return 0, err
 	}
 
 	// no committed randomness yet
-	if len(pubRandCommitMap) == 0 {
+	if pubRandCommit == nil || (pubRandCommit.StartHeight == 0 && pubRandCommit.NumPubRand == 0) {
```
Comment (gitferry) [inline finality-provider/service/fp_instance.go:900]: This would be more readable if we have a helper method of `PubRandCommit` like `pubRandCommit.IsEmpty()`

**Review by gitferry — APPROVED** [review]
Great work! Only nits

File: finality-provider/service/fp_instance.go (hunk 4 above)
Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:900]: I am a bit hesitant to create helper function b/c when it's empty it implies a bug in the consumer interface implementation. see conv here https://github.com/babylonchain/finality-provider/pull/438#discussion_r1657077298

File: clientcontroller/babylon/consumer.go (hunk 2 above)
Comment (bap2pecs) [reply clientcontroller/babylon/consumer.go:305]: ok added both to make it more readable:

```
	if len(res.PubRandCommitMap) == 0 {
		// expected when there is no PR commit at all
		return nil, nil
	}

	if len(res.PubRandCommitMap) > 1 {
		return nil, fmt.Errorf("expected length to be 1, but get :%d", len(res.PubRandCommitMap))
	}

	var commit *types.PubRandCommit = nil
	for height, commitRes := range res.PubRandCommitMap {
		commit = &types.PubRandCommit{
			StartHeight: height,
			NumPubRand:  commitRes.NumPubRand,
			Commitment:  commitRes.Commitment,
		}
	}

	return commit, nil
```

File: clientcontroller/cosmwasm/consumer.go (hunk 3 above)
Comment (bap2pecs) [reply clientcontroller/cosmwasm/consumer.go:430]: changed to 

```
	if len(commits) == 0 {
		// expected when there is no PR commit at all
		// `get_pub_rand_commit`'s return type is Vec<PubRandCommit> and it can be
		// empty vector if no results found
		return nil, nil
	}

	if len(commits) > 1 {
		return nil, fmt.Errorf("expected length to be 1, but got :%d", len(commits))
	}

	// Convert the response to the expected map format
	var commit *fptypes.PubRandCommit = nil
	for _, commitRes := range commits {
		commitCopy := commitRes // create a copy to avoid referencing the loop variable
		commit = &fptypes.PubRandCommit{
			StartHeight: commitCopy.StartHeight,
			NumPubRand:  commitCopy.NumPubRand,
			Commitment:  commitCopy.Commitment,
		}
	}

	return commit, nil
```

cc @maurolacy

File: finality-provider/service/fp_instance.go (hunk 4 above)
Comment (gitferry) [reply finality-provider/service/fp_instance.go:900]: If so, I think it makes more sense to move the check to the client controller implementation

Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:900]: maybe we can have sth like this:

```
func (prc *PubRandCommit) IsValid() bool {
    return prc.NumPubRand >= 1
}
```

wdyt? @gitferry

Comment (gitferry) [reply finality-provider/service/fp_instance.go:900]: Close to my point replied in https://github.com/babylonchain/finality-provider/pull/438#discussion_r1657169696. Maybe `Validate() error` is more informative

Comment (bap2pecs) [reply finality-provider/service/fp_instance.go:900]: > If so, I think it makes more sense to move the check to the client controller implementation

the problem I discussed w @SebastianElvis was the client impl can be incorrect and we want to add some check to detect it here

but I do see we can move it to `lastCommittedPublicRandWithRetry` at least

Comment (gitferry) [reply finality-provider/service/fp_instance.go:900]: Sounds good to me
